File: backend/app/schemas/lead.py
from enum import Enum
from typing import List, Optional
from datetime import datetime
import uuid
from pydantic import BaseModel, Field, ConfigDict

class LeadStatus(str, Enum):
    """
    Representa os estados possíveis de um Lead no sistema.
    """
    NOVO = "novo"
    EM_QUALIFICACAO = "em_qualificacao"
    QUALIFICADO = "qualificado"
    DESQUALIFICADO = "desqualificado"
    CONVERTIDO = "convertido"
    ARQUIVADO = "arquivado"

class LeadResponse(BaseModel):
    """
    Schema de resposta para detalhes de um Lead.
    """
    id: uuid.UUID
    nome: str
    email: Optional[str] = None
    telefone: str
    status: LeadStatus
    score: float = Field(default=0.0, description="Pontuação de qualificação do lead (0-100)")
    created_at: datetime
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "id": "550e8400-e29b-41d4-a716-446655440000",
                "nome": "João Silva",
                "email": "joao.silva@example.com",
                "telefone": "+5511999998888",
                "status": "novo",
                "score": 85.5,
                "created_at": "2024-05-12T10:00:00",
                "updated_at": "2024-05-12T10:30:00"
            }
        }
    )

class LeadsListResponse(BaseModel):
    """
    Schema de resposta para listagem de Leads com suporte a paginação.
    """
    items: List[LeadResponse]
    total: int = Field(..., description="Total de registros encontrados")
    page: int = Field(..., description="Página atual")
    pages: int = Field(..., description="Total de páginas disponível")

    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "items": [
                    {
                        "id": "550e8400-e29b-41d4-a716-446655440000",
                        "nome": "João Silva",
                        "email": "joao.silva@example.com",
                        "telefone": "+5511999998888",
                        "status": "novo",
                        "score": 85.5,
                        "created_at": "2024-05-12T10:00:00"
                    }
                ],
                "total": 1,
                "page": 1,
                "pages": 1
            }
        }
    )

File: frontend_flutter/lib/models/lead.dart
import 'package:equatable/equatable.dart';

class Lead extends Equatable {
  final String id;
  final String nome;
  final String? email;
  final String telefone;
  final String status;
  final double score;
  final DateTime criadoEm;

  const Lead({
    required this.id,
    required this.nome,
    this.email,
    required this.telefone,
    required this.status,
    required this.score,
    required this.criadoEm,
  });

  factory Lead.fromJson(Map<String, dynamic> json) {
    return Lead(
      id: json['id'] as String,
      nome: json['nome'] as String? ?? json['name'] as String? ?? '',
      email: json['email'] as String?,
      telefone: json['telefone'] as String? ?? json['phone'] as String? ?? '',
      status: json['status'] as String? ?? 'novo',
      score: (json['score'] as num?)?.toDouble() ?? 0.0,
      criadoEm: json['criado_em'] != null 
          ? DateTime.parse(json['criado_em'] as String)
          : (json['created_at'] != null 
              ? DateTime.parse(json['created_at'] as String)
              : DateTime.now()),
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'nome': nome,
    'email': email,
    'telefone': telefone,
    'status': status,
    'score': score,
    'criado_em': criadoEm.toIso8601String(),
  };

  @override
  List<Object?> get props => [id, nome, email, telefone, status, score, criadoEm];
}

class LeadsListResponse extends Equatable {
  final List<Lead> items;
  final int total;
  final int page;
  final int pages;

  const LeadsListResponse({
    required this.items,
    required this.total,
    required this.page,
    required this.pages,
  });

  factory LeadsListResponse.fromJson(Map<String, dynamic> json) {
    return LeadsListResponse(
      items: (json['items'] as List? ?? [])
          .map((e) => Lead.fromJson(e as Map<String, dynamic>))
          .toList(),
      total: json['total'] as int? ?? 0,
      page: json['page'] as int? ?? 1,
      pages: json['pages'] as int? ?? 1,
    );
  }

  @override
  List<Object?> get props => [items, total, page, pages];
}
